i18n-helpers/src/directives.rs

```diff
@@ -0,0 +1,130 @@
+use regex::{escape, Regex};
+
+#[derive(Debug, PartialEq)]
+pub enum Directive {
+    Skip,
+    TranslatorComment(String),
+}
+
+pub struct DirectiveFinder {
+    pattern_re: Regex,
+}
+
+impl Default for DirectiveFinder {
+    fn default() -> Self {
+        Self::new("i18")
+    }
+}
+
+impl DirectiveFinder {
+    pub fn new(prefix: &str) -> Self {
+        let pattern: String = r"<!-{2,}\s*".to_owned()
+            + &escape(prefix)
+            + &r"\s*[:-]"
+            + &r"(.*[^-])"  // the command part of the prefix, which must not end in a hyphen.
+            + &r"-{2,}>";
+        Self {
+            pattern_re: Regex::new(&pattern).expect("well-formed regex"),
+        }
+    }
+
+    pub fn find(&self, html: &str) -> Option<Directive> {
+        let captures = self.pattern_re.captures(html.trim())?;
+
+        let command = captures.get(1)?.as_str().trim();
+        match command.split(is_delimiter).next() {
+            Some("skip") => Some(Directive::Skip),
+            Some("comment") => {
```

Comment:
I think the regular expression could be made simpler if you write it out in full like this:

```
Regex::new(r"<!-{2,}\s*(i18n|mdbook-xgettext)\s*:\s*(skip|comment)\s*-{2,}>")
```

That is, hard-code the two options (they're linked directly to the enum) and also just hard-code the available prefixes.

I suggest that we clean up the mdbook-xgettext one as soon as we can and that simplifies things a bit more. In fact, we can clean it up right away: when we upgrade to the new version of `mdbook-i18n-helpers`, then we simply search-replace all of Comprehensive Rust in one atomic operation.

A small [search on GitHub](https://github.com/search?q=mdbook-xgettext+skip+path%3A*.md+NOT+repo%3Agoogle%2Fcomprehensive-rust+&type=code) shows that we're likely the only ones that use this feature right now, so this is another good reason to clean up any cruft right away.

Reply:
Acknowledged; I've applied the suggested hardcodings to simplify a bit.  I do think I want to keep the verbose regex (spread across a few lines with comments) because I'm finding it easier to read, especially because there's some subtlety toward the end of the regex.

Reply:
> I do think I want to keep the verbose regex (spread across a few lines with comments) because I'm finding it easier to read, especially because there's some subtlety toward the end of the regex.

Yeah, I'm fully supportive of using the verbose form! It's made exactly for complex cases like this :smile: